pynicotine/portmapper.py

```diff
@@ -126,17 +127,23 @@ def _request_port_mapping(self, public_port, private_port, lease_duration):
 
         with socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM, proto=socket.IPPROTO_UDP) as sock:
             sock.bind((self.local_ip_address, 0))
-            sock.settimeout(self.REQUEST_TIMEOUT)
-
             request = self.PortmapRequest(public_port, private_port, lease_duration)
-            request.sendto(sock, (self._gateway_address, self.REQUEST_PORT))
+            timeout = self.REQUEST_INIT_TIMEOUT
+            i = 1
 
-            try:
-                response = self.PortmapResponse(message=sock.recv(16))
-                return response.result
+            while i <= self.REQUEST_ATTEMPTS:
+                sock.settimeout(timeout)
+                request.sendto(sock, (self._gateway_address, self.REQUEST_PORT), i)
+
+                try:
+                    response = self.PortmapResponse(message=sock.recv(16))
+                    return response.result
+
+                except socket.timeout:
```

Comment:
Are exceptions other than `socket.timeout` possible?

If so, then there is a chance of an infinate loop here, because `i` will never increment in the case of some other error.

Reply:
You're right. I took care of it.

Reply:
An unhandled exception would break the loop.

Reply:
But I do agree that range() is the better option.

Reply:
How many total attempts does `self.REQUEST_ATTEMPTS + 1` result in?

Reply:
2